### ISSUE TITLE
Migrate from MSTest to xUnit.net/NUnit

### ISSUE THREAD

**Issue by hifi** [sshnet/SSH.NET#152]
This is a general proposal to move away from the proprietary MSTest suite to more modern open source and cross-platform unit testing facility.

Some points I gathered why this might be a good idea:

 - MSTest is no longer updated ([not entirely true, MSTest v2 exists](https://blogs.msdn.microsoft.com/visualstudioalm/2016/09/01/announcing-mstest-v2-framework-support-for-net-core-1-0-rtm/))
 - MSTest is closed source (could change with v2)
 - MSTest can't be used on non-Windows platforms (could change with v2)
 - xUnit.net/NUnit is probably easier to CI (more providers?)
 - **Even Microsoft is using xUnit.net in .NET Core**

However, there are some challenges:

 - There are *a lot* of existing tests to port
 - Are all current tests supported on xUnit.net/NUnit?
 - Can xUnit.net/NUnit target older frameworks that are supported by SSH.NET?

Implementation of xUnit.net/NUnit tests should probably be done side-by-side with the current MSTest tests and when they reach feature parity, MSTest tree would be removed from the repository. At this point there should already be a CI running xUnit.net/NUnit tests to make the switch seamless integration wise.

Also please note that I have not used *any* .NET/C# unit testing facility myself so I have no authoritative experience on any of them (though very limited use of NUnit) so everything in this proposal is purely based on my observations.

**Comment by drieseng** [comment]
MS has recently started working on [MSTest v2](https://blogs.msdn.microsoft.com/visualstudioalm/2016/09/01/announcing-mstest-v2-framework-support-for-net-core-1-0-rtm/) which supports .NET Core.
I'm also pretty sure it will be open-sourced in the coming months (but no, I have no inside info on this).

Which is to say, we _could_ consider sticking with MSTest.
If we decide to migrate away from MSTest, we should consider both xUnit and NUnit.
The community around NUnit is a lot more active.

Apart from the challenges you mentioned, support for older target frameworks can be a deciding factor.

**Comment by hifi** [comment]
Didn't stumble upon MSTest v2 when I looked for information, interesting. Still, one big reason why I would much prefer to move from MSTest to either of those (xUnit or NUnit) is that there's existing IDE integration for both xUnit and NUnit for Visual Studio *and* MonoDevelop/Xamarin Studio. It makes developing on non-Windows platforms unit testable *now* rather than some day in the future.

I did some testing of xUnit last night and it seemed to completely lack proper reference documentation but I did get nice IDE integration with MonoDevelop so that's something. NUnit would be the same regarding integration. I'll change the issue to be more generic regarding the choices.
